### PR TITLE
Avoid reducing unused Vector2 and Vector3 elements

### PR DESCRIPTION
**Pull request by tannergooding**
Reduce only the meaningful `Vector2` and `Vector3` elements, avoiding zero-extension and extra arithmetic from four-element reductions. Keep intermediates in SIMD form, with scalar reductions for dot products, lengths, and distances, and vector results for normalization and reflection. The implementations use portable `Vector128` operations, without architecture-specific paths.

This also preserves negative zero when all summed elements are negative zero, rather than adding an irrelevant positive zero. Adds coverage for signed zero, grouping, nonfinite inputs, and unused upper elements.

Addresses the `Vector2`/`Vector3` reduction portion of dotnet/runtime#133297. `Vector4`, JIT simplifications, and loop alignment are out of scope.

----------

Release x64 codegen for the local `VectorBench` loops, compared with the original upstream implementation:

| Method | Before | After |
|---|---:|---:|
| `LengthSquared2` | 66 bytes | 63 bytes |
| `LengthSquared3` | 73 bytes | 67 bytes |
| `Normalize3` | 95 bytes | 94 bytes |
| `Reflect3` | 117 bytes | 116 bytes |

These are whole-method sizes including alignment; the `LengthSquared2` loop itself shrinks from 46 to 30 bytes.

The portable `Vector3` broadcast costs five instructions / 25 reduction bytes versus four / 18 for an investigated SSE specialization. In local Ryzen 9 7950X `VectorBench` loops over 1,000 vectors, `Normalize3` was within 1% faster and `Reflect3` was 11-12% slower than that specialization in both runtime orders. This comparison is against the rejected specialization, not upstream; portability is preferred over separate target-specific reduction paths. ARM and WASM performance has not been measured.

> [!NOTE]
> This description was drafted with GitHub Copilot.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs [hunk 1]
```diff
@@ -964,7 +976,20 @@ public static (Vector3 Sin, Vector3 Cos) SinCos(Vector3 vector)
...
+            return (Vector128.Shuffle(vector, Vector128.Create(2)) + sum).ToScalar();
+        }
+
+        [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static Vector3 Sum(Vector128<float> value)
+        {
+            Vector128<float> sum = Vector128.Shuffle(value, Vector128.Create(1, 0, 3, 2)) + value;
+            return Vector128.Create((Vector128.Shuffle(value, Vector128.Create(2)) + sum).ToScalar()).AsVector3();
+        }
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs]: Renamed to `SumAndBroadcast` to make the replicated result explicit.

> [!NOTE]
> This reply was drafted with GitHub Copilot.

File: src/libraries/System.Numerics.Vectors/tests/Vector3Tests.cs [hunk 2]
```diff
@@ -1845,9 +1845,73 @@ public void ShuffleTest(float x, float y, float z)
...
+
+            Assert.Equal(11.0f, Vector3.Sum(value));
+            Assert.Equal(49.0f, Vector3.Dot(value, value));
+            Assert.Equal(7.0f, value.Length());
+            Assert.Equal(49.0f, value.LengthSquared());
+            Assert.Equal(7.0f, Vector3.Distance(value, Vector3.Zero));
+            Assert.Equal(49.0f, Vector3.DistanceSquared(value, Vector3.Zero));
+            Assert.Equal(Vector3.Create(2.0f / 7.0f, 3.0f / 7.0f, 6.0f / 7.0f), Vector3.Normalize(value));
+            Assert.Equal(Vector3.Create(-2.0f, 3.0f, 6.0f), Vector3.Reflect(value, normal));
         }
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/libraries/System.Numerics.Vectors/tests/Vector3Tests.cs:1923]: Added a comment clarifying the limitation. `AsVector3` is intrinsic: in the actual Release x64 test codegen, the supplied value remains in lane 3 and feeds the inlined reductions. The test exercises a poisoned lane when retained, but does not assume that preservation is guaranteed across backends.

> [!NOTE]
> This reply was drafted with GitHub Copilot.

File: src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs [hunk 3]
```diff
@@ -1816,9 +1816,42 @@ public void ShuffleTest(float x, float y)
...
+
+            Assert.Equal(7.0f, Vector2.Sum(value));
+            Assert.Equal(25.0f, Vector2.Dot(value, value));
+            Assert.Equal(5.0f, value.Length());
+            Assert.Equal(25.0f, value.LengthSquared());
+            Assert.Equal(5.0f, Vector2.Distance(value, Vector2.Zero));
+            Assert.Equal(25.0f, Vector2.DistanceSquared(value, Vector2.Zero));
+            Assert.Equal(Vector2.Create(3.0f / 5.0f, 4.0f / 5.0f), Vector2.Normalize(value));
+            Assert.Equal(Vector2.Create(-3.0f, 4.0f), Vector2.Reflect(value, normal));
         }
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs:1889]: Added a comment clarifying the limitation. `AsVector2` is intrinsic: in the actual Release x64 test codegen, the supplied values remain in lanes 2/3 and feed the inlined reductions. The test exercises poisoned lanes when retained, but does not assume that preservation is guaranteed across backends.

> [!NOTE]
> This reply was drafted with GitHub Copilot.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs [hunk 4]
```diff
@@ -926,7 +938,14 @@ public static (Vector2 Sin, Vector2 Cos) SinCos(Vector2 vector)
...
         [Intrinsic]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static float Sum(Vector2 value) => Vector128.Sum(value.AsVector128());
+        public static float Sum(Vector2 value) => Sum(value.AsVector128Unsafe()).ToScalar();
+
+        [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static Vector2 Sum(Vector128<float> value)
+        {
+            return (value + Vector128.Shuffle(value, Vector128.Create(1, 0, 3, 2))).AsVector2();
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:947]: Renamed to `SumAndBroadcast` to make the replicated result explicit.

> [!NOTE]
> This reply was drafted with GitHub Copilot.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs [hunk 5]
```diff
@@ -926,7 +938,14 @@ public static (Vector2 Sin, Vector2 Cos) SinCos(Vector2 vector)
...
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static float Sum(Vector2 value) => Vector128.Sum(value.AsVector128());
+        public static float Sum(Vector2 value) => SumAndBroadcast(value.AsVector128Unsafe()).ToScalar();
+
+        [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static Vector2 SumAndBroadcast(Vector128<float> value)
+        {
+            return (value + Vector128.Shuffle(value, Vector128.Create(1, 0, 3, 2))).AsVector2();
+        }
```
[the remark this answers is not in the document]
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:948]: `NaNs` are not guaranteed like that, never have been.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs [hunk 6]
```diff
@@ -1104,15 +1123,23 @@ public readonly bool TryCopyTo(Span<float> destination)
         /// <altmember cref="LengthSquared" />
         [Intrinsic]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public readonly float Length() => Vector128.Length(this.AsVector128());
+        public readonly float Length()
+        {
+            Vector128<float> vector = this.AsVector128Unsafe();
+            return float.Sqrt(Sum((vector * vector).AsVector2()));
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:1129]: Took me a while to parse the `sqrt` here, but I take it `Length()` in this context denotes the 2-norm of the vector, right?

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs [hunk 7]
```diff
@@ -1816,9 +1816,44 @@ public void ShuffleTest(float x, float y)
...
+                Assert.Equal(BitConverter.SingleToInt32Bits(expectedResult), BitConverter.SingleToInt32Bits(dot[i]));
+            }
+        }
+
+        [Theory]
+        [InlineData(float.NaN)]
+        [InlineData(float.PositiveInfinity)]
+        [InlineData(float.NegativeInfinity)]
+        [InlineData(42.0f)]
+        public void ReductionsIgnoreUpperElementsTest(float upper)
```
Comment (krwq) [inline src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs:1874]: > [!NOTE]
> This comment was generated with the help of AI. Please review before applying.

Vector2 is missing the `ReductionsWithNonFiniteElementsTest` counterpart that Vector3 gets — the new NaN / +Inf / -Inf coverage lives only on the Vector3 side. The two Vector2 tests here cover signed zeros (`SumAndDotTest`) and poisoned upper lanes (this one), but never seed NaN or +/-Inf into lane 0 or 1 of a Vector2 and assert that `Vector2.Sum`, `Vector2.Dot`, `Vector2.Length`, `Vector2.Distance`, `Vector2.Normalize`, and `Vector2.Reflect` propagate them correctly.

**Why it matters.** Vector2's reductions changed too — `Reflect` and `Normalize` now use `AsVector128Unsafe` + `SumAndBroadcast` instead of `Vector128.Dot`/`Vector128.Normalize`. Locking down the NaN/Inf propagation contract on the 2-element side (as this PR did on the 3-element side) would catch a future regression where, for instance, a JIT intrinsic recognizer quietly drops a term in `Vector2.Reflect` while leaving Vector3 untouched.

**Suggested change.** Add a `ReductionsWithNonFiniteElementsTest` for Vector2 shaped like the Vector3 one — same inline data pattern (NaN in x, NaN in y, +/-Inf, +Inf paired with -Inf), asserting `Sum`, `Dot`, `Normalize`, and `Reflect` each match the scalar formula for every lane.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs [hunk 8]
```diff
@@ -964,7 +976,20 @@ public static (Vector3 Sin, Vector3 Cos) SinCos(Vector3 vector)
         /// <inheritdoc cref="Vector4.Sum(Vector4)" />
         [Intrinsic]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static float Sum(Vector3 value) => Vector128.Sum(value.AsVector128());
+        public static float Sum(Vector3 value)
```
Comment (krwq) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs:979]: > [!NOTE]
> This comment was generated with the help of AI. Please review before applying.

`Vector3.Sum` inlines the same two-shuffle reduction that `Vector3.SumAndBroadcast` (a few lines below) already contains, whereas `Vector2.Sum` cleanly delegates: `SumAndBroadcast(value.AsVector128Unsafe()).ToScalar()`. Same reduction expression here (`Shuffle(v,{1,0,3,2}) + v`, then `Shuffle(v,{2}) + sum`, then `ToScalar`), then `SumAndBroadcast` immediately re-broadcasts that scalar — so the extra copy adds nothing.

**Why it matters.** Two copies of the same 3-lane reduction on a hot path are a maintenance hazard. A future correctness fix (reduction order, a different shuffle pattern) has to be applied in both places or the two drift apart silently.

**Suggested change.**

```csharp
public static float Sum(Vector3 value) => SumAndBroadcast(value.AsVector128Unsafe()).ToScalar();
```

The chain `Vector128.Create(scalar).AsVector3().AsVector128Unsafe().ToScalar()` is the same broadcast+extract round trip `Vector2.Sum` already relies on, and the JIT collapses it in that path. If `VectorBench` shows a codegen regression at `[MethodImpl(AggressiveInlining)]` scope, drop the suggestion — worth confirming against the same disassembly numbers cited in the PR description before landing the refactor.

**Review by krwq — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs (hunk 6 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:1129]: It's the Euclidean norm, yes (aka magnitude). `Length` is the standard term for it in this particular space.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs (hunk 8 above)
Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector3.cs:979]: This differs to avoid the extra inlining step on a more perf critical type.

File: src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs (hunk 7 above)
Comment (tannergooding) [reply src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs:1874]: Vector3 has it because its the more interesting type due to covering 3/4ths, rather than half, of the vector. So it is less trivial to filter out the irrelevant parts.

Comment (tannergooding) [reply src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs:1874]: Added this coverage since I needed to skip a test on Mono

**Comment by tannergooding** [comment]
@eiriktsarpalis, @krwq could this get a fresh sign-off? I pushed the parity test coverage for Vector2 and added skip for a scenario on Mono involving a `-0` edge case. It's not worth the extra churn in Mono to fix since that should be getting removed later here this release anyways.

**Review by eiriktsarpalis — APPROVED** [review]

**Comment by tannergooding** [comment]
/ba-g remaining failures are unrelated in mono and also occurring in main

**Comment by tannergooding** [comment]
/backport to release/11.0

> [!NOTE]
> This backport request was posted with GitHub Copilot.
